lib/galaxy/tools/toolbox/lineages/factory.py

```diff
@@ -20,28 +11,39 @@ def __init__(self, app):
         self.lineage_map = {}
         self.app = app
 
-    def register(self, tool, from_toolshed=False):
+    def register(self, tool):
         tool_id = tool.id
-        versionless_tool_id = remove_version_from_guid( tool_id )
-        if from_toolshed:
-            lineage = ToolShedLineage.from_tool(self.app, tool)
-        else:
+        versionless_tool_id = remove_version_from_guid( tool_id)
+        lineage = self.lineage_map.get(versionless_tool_id)
+        if not lineage:
             lineage = StockLineage.from_tool( tool )
         if versionless_tool_id and versionless_tool_id not in self.lineage_map:
             self.lineage_map[versionless_tool_id] = lineage
         if tool_id not in self.lineage_map:
             self.lineage_map[tool_id] = lineage
+        lineage.register_version(tool.version)
```

Comment:
OK, I think this is what happens:
When a tool is loaded the first time, it will enter `self.lineage_map`, and have its version registered.
When a related tool is added, it will be fetched from `self.lineage_map` via the `versionless_tool_id`,  and not run through `ToolLineage.from_tool`. So we need to add the current tool's version via `lineage.register_version(tool.version)`. I'll add a comment to clarify this.